Refactor and Test onCreatePage

jestapinski opened the pull request:
This PR does a few things:

- Refactors logic to generate the filters for the learn page to not repeatedly call stitch. Instead we use the article objects we have to update language/product values.
- Cleans up var names from a prior PR
- Adds testing for both generating the filters and getting featured articles
- Adds a `utils` subdirectory in `/tests` with a simple config (since this is mainly for js helpers, not components).

jestapinski commented on src/utils/setup/on-create-page.js:
```diff
@@ -56,54 +56,57 @@ const getFeaturedArticles = (allArticles, featuredArticleSlugs) => {
+            });
+        }
+        // Go through products for this article and update filter info
+        if (productsForArticle) {
```
There is some logical redundancy between this conditional and the one on line 67, and so I tried to refactor into a single helper. This got quite complicated and much harder to understand, so I left these separate.

jbouzi12 replied:
Yea I see how these are very similar but the fields within the resulting data structures are where things can get hairy when trying to combine these. Is there a way we can unify the naming for the resulting objects in a way that makes sense and allow us to create a single helper?

jestapinski replied:
That's a good point, perhaps something like:
```
languages: {
 count: 1,
 otherTagCounts: {
   stitch: 1,
 }
}
```

The biggest con here is this is clear now, but if we add a third filter it would be a bit harder to understand. I think there might be a way to ratify some of the difference here just not sure what the best way forward is, any thoughts?

jbouzi12 replied:
Yea that makes sense. Hmm I'll have to think about this a little more, especially considering the additional filters the scenario you raised. I'm ok with this moving forward for now though.

greggb replied:
I'm happy with whatever direction y'all decide, but I'm okay with a bit of redundancy for clarity/extensibility here. As long as it's still straightforward to test.

jestapinski replied:
I think at this point having the distinction between languages/products and products/languages supports keeping these a bit separate to understand. We can revisit if we find a great way to refactor, but I think this is the best balance of code clarity and testibility for now

greggb commented on src/utils/setup/on-create-page.js:
```diff
@@ -56,54 +56,57 @@ const getFeaturedArticles = (allArticles, featuredArticleSlugs) => {
-    }
-    return filters;
+    allArticles.forEach(a => {
+        const languagesForArticle = a.query_fields.languages;
```
I think it'd be better to expand these single letter vars
a => article
l => language or lang
p => product

greggb commented on src/utils/setup/on-create-page.js:
```diff
@@ -56,54 +56,57 @@ const getFeaturedArticles = (allArticles, featuredArticleSlugs) => {
+    allArticles.forEach(a => {
+        const languagesForArticle = a.query_fields.languages;
+        const productsForArticle = a.query_fields.products;
+        // Go through languages for this article and update filter info
```
Let's add a little more to this like what filter info we're including (count etc)

greggb commented on tests/utils/get-featured-articles.test.js:
```diff
@@ -0,0 +1,45 @@
+import { getFeaturedArticles } from '../../src/utils/setup/on-create-page';
+
+it('should correctly find featured articles given a set of requested articles', () => {
```
Should we think about capping the list to 4 and 3 for home and learn too?

jestapinski replied:
Yes we definitely should. The home page is flexible to support <= 4, but the learn page should really be 3 (otherwise it would break the design?).

greggb commented on src/utils/setup/on-create-page.js:
```diff
@@ -56,54 +56,57 @@ const getFeaturedArticles = (allArticles, featuredArticleSlugs) => {
+            });
+        }
+    });
+    return { languages, products };
```
👍

greggb commented on tests/utils/get-learn-page-filters.test.js:
```diff
@@ -0,0 +1,85 @@
+        },
+    ];
+
+    const result = getLearnPageFilters(allArticles);
```
```suggestion
    const filterData = getLearnPageFilters(allArticles);
```

greggb commented on tests/utils/get-learn-page-filters.test.js:
```diff
@@ -0,0 +1,85 @@
+
+    const result = getLearnPageFilters(allArticles);
+    const languages = result.languages;
+    const products = result.products;
```
nit suggestion - check for result.language/products right in the expect rather than assigning. It'll draw a more clear line from the work to the test. I'd also move the `result` assignment right above the `expect`s

jestapinski commented:
> Looks great - really nice refactor. Any stats on the build time increase/decrease after this? I had some suggestions on code style that might help.

Down by about ~2 seconds last time I checked. I'll check again after addressing the comments

jestapinski commented on src/utils/setup/on-create-page.js:
```diff
@@ -50,54 +54,67 @@ const findArticlesFromSlugs = (allArticles, articleSlugs) => {
+                }
+                if (productsForArticle) {
+                    // If this article also has products, update those counts as well for this language
+                    const productsForLanguage = languages[language].products;
```
@greggb interesting note on these vars. Since we are updating destructively we have to leave off the last key to allow the operation to modify the referenced object and not just the scoped variable :)

jestapinski commented:
Thanks all!
